### PR TITLE
PHPC-640: Implement interfaces for userland BSON type classes

### PR DESCRIPTION
**Pull request by jmikola**
https://jira.mongodb.org/browse/PHPC-640
https://jira.mongodb.org/browse/PHPC-937

Related:

 * https://github.com/mongodb/mongo-hhvm-driver/blob/master/serialization.rst#scalar-types
 * https://github.com/mongodb/mongo-hhvm-driver/pull/111

### CODE REVIEW

File: php_bson.h [hunk 1]
```diff
@@ -60,11 +69,7 @@ typedef struct {
 	bool                     is_visiting_array;
 } php_phongo_bson_state;
 
-#if PHP_VERSION_ID >= 70000
-#define PHONGO_BSON_STATE_INITIALIZER  { {{ 0 }}, { PHONGO_TYPEMAP_NONE, NULL, PHONGO_TYPEMAP_NONE, NULL, PHONGO_TYPEMAP_NONE, NULL}, NULL, 0 }
-#else
-#define PHONGO_BSON_STATE_INITIALIZER  { NULL, { PHONGO_TYPEMAP_NONE, NULL, PHONGO_TYPEMAP_NONE, NULL, PHONGO_TYPEMAP_NONE, NULL}, NULL, 0 }
-#endif
+#define PHONGO_BSON_STATE_INITIALIZER  { 0 }
```
Comment (derickr) [inline php_bson.h]: I don't believe this is right. I had some compilers whine about this change to just have a single { 0 } in Xdebug: https://github.com/xdebug/xdebug/commit/7ad9db738bfa80d69621e4c7f66a04cd361269f9

File: php_phongo_classes.h [hunk 2]
```diff
@@ -242,18 +242,19 @@ extern zend_class_entry *php_phongo_connectiontimeoutexception_ce;
...
-extern zend_class_entry *php_phongo_persistable_ce;
-extern zend_class_entry *php_phongo_unserializable_ce;
-extern zend_class_entry *php_phongo_serializable_ce;
 extern zend_class_entry *php_phongo_binary_ce;
 extern zend_class_entry *php_phongo_decimal128_ce;
 extern zend_class_entry *php_phongo_javascript_ce;
 extern zend_class_entry *php_phongo_maxkey_ce;
 extern zend_class_entry *php_phongo_minkey_ce;
 extern zend_class_entry *php_phongo_objectid_ce;
+extern zend_class_entry *php_phongo_persistable_ce;
```
Comment (derickr) [inline php_phongo_classes.h]: I don't think it makes sense to reorganise this. The specific classes' ce's should be together, but not mixed in with their interfaces. The rest of the CE's isn't alphabetised either.

File: src/bson-encode.c [hunk 3]
```diff
@@ -113,7 +116,32 @@ static int php_phongo_is_array_or_document(zval *val TSRMLS_DC) /* {{{ */
...
+		zend_call_method_with_0_params(object, NULL, NULL, "toBSONType", &retval);
+#else
+		zval *retval;
+
+		zend_call_method_with_0_params(&object, NULL, NULL, "toBSONType", &retval);
+#endif
+
+		if (Z_ISUNDEF(retval)) {
+			/* zend_call_method() failed */
+			return;
```
Comment (derickr) [inline src/bson-encode.c]: Shouldn't that throw an exception?

File: src/bson.c [hunk 4]
```diff
@@ -35,6 +35,30 @@
...
+# define PHONGO_BSON_STATE_ZCHILD(state) (&((php_phongo_bson_state *)(state))->zchild)
+#else
+# define PHONGO_BSON_STATE_ZCHILD(state) (((php_phongo_bson_state *)(state))->zchild)
+#endif
+
+#if PHP_VERSION_ID >= 70000
+# define PHONGO_BSON_TYPE_WRAPPER(wrapper_ce, zchild) \
+	if (wrapper_ce) { \
+		zval zwrapper; \
+		zend_call_method_with_1_params(NULL, wrapper_ce, NULL, "createfrombsontype", &zwrapper, &(zchild)); \
```
Comment (derickr) [inline src/bson.c]: Do you not have a constant for "createfrombsontype" ?

File: src/bson-encode.c (hunk 3 above)
Comment (jmikola) [inline src/bson-encode.c]: [`zend_call_method()`](https://github.com/php/php-src/blob/PHP-7.1.4/Zend/zend_interfaces.c#L34) uses `zend_error_noreturn()`. I suppose PHP 7 might throw an Error object from that, but I expect basic error reporting in PHP 5.x.

This is consistent with other handling we do for potential `zend_call_method()` failures (e.g. `bsonSerialize()`).

File: src/bson.c (hunk 4 above)
Comment (jmikola) [inline src/bson.c]: I did not create a constant like we had for `BSON_UNSERIALIZE_FUNC_NAME`, as I recall some case where `zend_call_method()` could not find the function when using `"createFromBSONType"`. I wasn't able to determine at the time why that wasn't an issue for `BSON_UNSERIALIZE_FUNC_NAME` (defined as `"bsonUnserialize"`), but I'll check again.

File: php_bson.h (hunk 1 above)
Comment (jmikola) [inline php_bson.h]: Do you recall what compilers were affected (e.g. clang, MS)?

I can always revert this, but I'd be curious to test it.

File: php_phongo_classes.h (hunk 2 above)
Comment (jmikola) [inline php_phongo_classes.h]: The `MongoDB\Driver\Exception` CEs are not alphabetized, but `MongoDB\Driver` and `MongoDB\Driver\Monitoring` CEs were, as were all externs for the init functions. I'll revert this change, though.

File: src/bson.c (hunk 4 above)
Comment (jmikola) [inline src/bson.c]: Confirmed that `zend_call_method()` allows `function_name` to have any case when invoking a method on an object (specifying the first `obj` arg); however, invoking a static method when only supplying `obj_ce` requires the function name to be lowercase. I assume this has something to do with differences in storing the function entries on the CE and object instances.

In any event, these strings are only used twice in these two macros, unlike the `bsonSerialize` and `bsonUnserialize` constants, so I will not create a new constant.

File: php_bson.h (hunk 1 above)
Comment (jmikola) [inline php_bson.h]: Note that we're already using `{0}` for initializing potentially nested structs (e.g. `smart_str` in PHP 7: https://github.com/mongodb/mongo-php-driver/blob/608ce77839e1760f9a7272aa61d1158d6bcd741c/src/BSON/ObjectID.c#L199) and we've received no complaints about build issues.

http://stackoverflow.com/a/1761873/162228 suggests that this is kosher.

**Pull request merged**